src/Microsoft.AspNet.Mvc.Razor/Compilation/ExpressionRewriter.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNet.Mvc.Razor.Compilation
+{
+    /// <summary>
+    /// An expression rewriter which can hoist a simple expression lambda into a private field.
+    /// </summary>
+    public class ExpressionRewriter : CSharpSyntaxRewriter
+    {
+        private static readonly string FieldNameTemplate = "__h{0}";
+
+        public ExpressionRewriter(SemanticModel semanticModel)
+        {
+            SemanticModel = semanticModel;
+
+            Expressions = new List<KeyValuePair<SimpleLambdaExpressionSyntax, IdentifierNameSyntax>>();
+        }
+
+        private SemanticModel SemanticModel { get; }
+
+        private List<KeyValuePair<SimpleLambdaExpressionSyntax, IdentifierNameSyntax>> Expressions { get; }
+
+        public override SyntaxNode VisitClassDeclaration(ClassDeclarationSyntax node)
+        {
+            Expressions.Clear();
+
+            // Call base first to visit all the children and populate Expressions.
+            var classDeclaration = (ClassDeclarationSyntax)base.VisitClassDeclaration(node);
+
+            var memberDeclarations = new List<MemberDeclarationSyntax>();
+            foreach (var kvp in Expressions)
+            {
+                var expression = kvp.Key;
+                var memberName = kvp.Value.GetFirstToken();
+
+                var expressionType = SemanticModel.GetTypeInfo(expression);
+                var declaration = SyntaxFactory.FieldDeclaration(
+                    SyntaxFactory.List<AttributeListSyntax>(),
+                    SyntaxFactory.TokenList(
+                        SyntaxFactory.Token(SyntaxKind.PrivateKeyword),
+                        SyntaxFactory.Token(SyntaxKind.StaticKeyword),
+                        SyntaxFactory.Token(SyntaxKind.ReadOnlyKeyword)),
+                    SyntaxFactory.VariableDeclaration(
+                        SyntaxFactory.ParseTypeName(expressionType.ConvertedType.ToDisplayString()),
```

Comment:
Looks like we should be passing https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayFormat.cs#L125 to include `global::` but otherwise seems right.